src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py

```diff
@@ -0,0 +1,121 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Formula generator from component graph for Consumer Power."""
+
+from __future__ import annotations
+
+from collections import abc
+
+from ....microgrid import connection_manager
+from ....microgrid.component import Component, ComponentCategory, ComponentMetricId
+from .._formula_engine import FormulaEngine
+from .._resampled_formula_builder import ResampledFormulaBuilder
+from ._formula_generator import ComponentNotFound, FormulaGenerator
+
+
+class ConsumerPowerFormula(FormulaGenerator):
+    """Formula generator from component graph for calculating the Consumer Power."""
+
+    def generate(self) -> FormulaEngine:
+        """Generate formula for calculating consumer power from the component graph.
+
+        Returns:
+            A formula engine that will calculate the consumer power.
+
+        Raises:
+            ComponentNotFound: If the component graph does not contain a consumer power
+                component.
+        """
+        builder = self._get_builder("consumer-power", ComponentMetricId.ACTIVE_POWER)
+        component_graph = connection_manager.get().component_graph
+        grid_component = next(
+            (
+                comp
+                for comp in component_graph.components()
+                if comp.category == ComponentCategory.GRID
+            ),
+            None,
+        )
+
+        if grid_component is None:
+            raise ComponentNotFound("Grid component not found in the component graph.")
+
+        grid_successors = component_graph.successors(grid_component.component_id)
+        if not grid_successors:
+            raise ComponentNotFound("No components found in the component graph.")
+
+        if len(grid_successors) == 1:
+            grid_meter = next(iter(grid_successors))
+            assert (
+                grid_meter.category == ComponentCategory.METER
+            ), "Only grid successor is not a meter."
+            return self._gen_with_grid_meter(builder, grid_meter)
+        return self._gen_without_grid_meter(builder, grid_successors)
+
+    def _gen_with_grid_meter(
+        self,
+        builder: ResampledFormulaBuilder,
+        grid_meter: Component,
+    ) -> FormulaEngine:
+        """Generate formula for calculating consumer power with grid-side meter.
```

Comment:
We should define what a grid side meter is.

Reply:
maybe just calling it the grid meter makes it more obvious?

Reply:
renamed to "grid meter" in the docstrings.